### PR TITLE
feat: add exit to empty seat

### LINKED ISSUE #1582: Allow Empty Seat to Have a Proposal with Want, Exit
**What is the Problem Being Solved?**
https://github.com/Agoric/agoric-sdk/pull/1389 solved the problem of needing to add a new seat, synchronously, from the contract side. However, it did so in a way that made this type of seat different from other seats in two related ways: there was no proposal, so there could never be any offer-safety enforced, and there was no exit rule specified in the proposal, so there could never be an exit `afterDeadline`.

However, neither of these restrictions is necessary. It's really only the escrowing of payments that is difficult, since the challenge is to do escrowing from the ZCF side through Zoe in a way compatible with getting a seat synchronously. But, we can add `want` and `exit` rules without trying to escrow anything and thus avoiding these difficulties.

**Description of the Design**

Allow calls to `zcf.makeEmptySeatKit` to pass in two arguments: `want` and `exit` which will be used to make a proposal. The exitObj is made like any other seat, meaning that the userSeat can be used to exit in exactly the same ways as other seats. If `onDemand` is specified as the exit rule, the seat can be exited on demand using the userSeat specifically. If `afterDeadline` is chosen, the seat will be exited after the deadline. If `waived` is chosen, the userSeat cannot exit. Of course, the zcfSeat can exit, but if the userSeat can be passed on to someone else or a different contract, it's still valuable to make the distinction in authority.

**~~Security~~ Considerations**

I don't know whether this counts as a security consideration per se, but currently, under offer safety, giving nothing means getting nothing back fulfills the refund requirement. So specifying a `want` without any `give` is basically useless. It seems like it would be very useful to be able to make empty seats that have a `want`, but which give nothing, and which have offer safety *fail* if `want` is not fulfilled. I'm thinking in particular of ongoing contracts with a manager who is conducting many sales. For each of those sales, it'd be nice to make a new seat to represent the offer safety from the selling side, taking inventory from another, long-lasting seat. Perhaps rather than escrowing in the `zcf.makeEmptySeatKit` call, we should be doing a reallocate, such that `give` in the proposal for that seat makes sense and only has to match the reallocated amount, with no payments involved.

**Test Plan**
Add tests to exercise the userSeat methods to make sure they work the same as all other userSeats.

### PR DESCRIPTION
**Pull request by katelynsills**
See #1582 for more discussion. I'm curious if we could implement the `give` portion too, using a reallocation.

I still need to add tests of the different userSeat methods to ensure the empty seat operates like a real seat in all ways.

Closes #1582

### CODE REVIEW

**Review by Chris-Hibbert — COMMENTED** [review]
This looks like it continues to conserve all our safety properties, but I don't understand its value. 

Under what circumstances would a contract want to create an internal seat for which it needed an exit clause?

**Comment by katelynsills** [comment]
> This looks like it continues to conserve all our safety properties, but I don't understand its value.
> 
> Under what circumstances would a contract want to create an internal seat for which it needed an exit clause?

Many circumstances - for instance, if the seat should exit after a particular deadline. Or, if the userSeat facet will be handed to someone else such that the ability to exit is separated from the contract's ability to exit.

My point, though, is that the burden of proof has been in the wrong direction. We shouldn't be asking why seats should all act the same. We should be asking why some seats should have different functionality. I would like the empty seats to have the same functionality as other seats.

**Review by Chris-Hibbert — APPROVED** [review]

**Review by erights — COMMENTED** [review]
I like the switching burden of proof argument. But it seems this PR is only a step along the way. The result is far from like a seat in all ways.

A seat whose `want` alone must be satisfied to preserve offer safety would take it farther from being just like a normal seat. I also don't see how it is a workable plan. The seat would start with an allocation that ***VIOLATES*** its offer safety. If the contract crashes in this state, offer safety would be violated. If the seat is exited in this state, what happens? Preventing the exit violates payout liveness.

So putting aside the burden-of-proof argument, it would help to actually have and try a concrete use case. The issues under even an artificial but concrete use case may be much easier to evaluate than under imagined use.

Was there a particular use case you had in mind that motivated this in the first place?

**Comment by katelynsills** [comment]
> I like the switching burden of proof argument. But it seems this PR is only a step along the way. The result is far from like a seat in all ways.

I think I disagree with this. The only way it is different is the offer cannot `give` anything. That's only one difference. An important difference, but it is the only one. 

I view this PR as a step towards being able to `give` from the contract side. Potential solutions include either 1) being able to synchronously process payments from the contract to cover `give` (hard) or 2) being able to synchronously use reallocations to cover `give` (seems easier). Either way, this PR shows that our previous assumption that you cannot exit an empty seat was unnecessary. 

> A seat whose `want` alone must be satisfied to preserve offer safety would take it farther from being just like a normal seat. I also don't see how it is a workable plan. The seat would start with an allocation that _**VIOLATES**_ its offer safety. If the contract crashes in this state, offer safety would be violated. If the seat is exited in this state, what happens? Preventing the exit violates payout liveness.

Yup, that makes sense.

> So putting aside the burden-of-proof argument, it would help to actually have and try a concrete use case. The issues under even an artificial but concrete use case may be much easier to evaluate than under imagined use.
> Was there a particular use case you had in mind that motivated this in the first place?

Yes, 1) the ability to exit an empty seat automatically `afterDeadline` and 2) to hand the userSeat to another entity such that they (who do not have access to the zcf seat) can exit on demand.

I think I will try to change this PR to allow the contract code to pass *only* the exit rule, so that the `want` and `give` are still empty under the hood. I want to push this PR forward because I think it does two important things: 1) allows all Seats to have uniform exitObjs based on the exit rule, and 2) shows how the proposal can be updated from the ZCF side, preparing the way in the future for us handling `give` correctly.

**Comment by erights** [comment]
> shows how the proposal can be updated from the ZCF side, preparing the way in the future for us handling give correctly.

By "updated" do you mean modified after creation? I don't think so, and your current code only provides proposal ingredients when the seat is constructed. This makes sense to me.

With that clarified, I look forward to the next version. Thanks.

**Comment by katelynsills** [comment]
> By "updated" do you mean modified after creation? I don't think so, and your current code only provides proposal ingredients when the seat is constructed. This makes sense to me.

No, sorry, updated is the wrong word. I mean created.

File: packages/zoe/src/types.js [hunk 1]
```diff
@@ -208,7 +208,7 @@
  * @property {(brand: Brand) => Issuer} getIssuerForBrand
  * @property {GetAmountMath} getAmountMath
  * @property {MakeZCFMint} makeZCFMint
- * @property {() => ZcfSeatKit} makeEmptySeatKit
+ * @property {(want: AmountKeywordRecord=, exit: ExitRule= }) => ZcfSeatKit} makeEmptySeatKit
```
Comment (erights) [inline packages/zoe/src/types.js]: Stale `want:`. Please remove.

File: packages/zoe/src/zoeService/zoe.js [hunk 2]
```diff
@@ -385,10 +386,8 @@ function makeZoe(vatAdminSvc, zcfBundleName = undefined) {
             instanceAdmin,
             proposal,
             brandToPurse,
-            {
-              offerResult: offerResultPromiseKit.promise,
-              exitObj: exitObjPromiseKit.promise,
-            },
+            exitObjPromiseKit.promise,
```
Comment (erights) [inline packages/zoe/src/zoeService/zoe.js:403]: Didn't you already flatten this out in a previous PR? In any case, the flattening is good.

**Review by erights — APPROVED** [review]
LGTM

File: packages/zoe/src/zoeService/zoe.js (hunk 2 above)
Comment (katelynsills) [reply packages/zoe/src/zoeService/zoe.js:403]: Whoops, I've got some rebasing to do, sorry.

File: packages/zoe/src/types.js (hunk 1 above)
Comment (katelynsills) [reply packages/zoe/src/types.js]: Good point
